### PR TITLE
WebSocket implementation

### PR DESCRIPTION
**Pull request by 4e6**
Changelog:
- add: implement WebSocket
- add: tests for timers and crypto components
- update: the demo program to check that we can work with Yjs as a client

### CODE REVIEW

File: src/main/java/module-info.java [hunk 1]
```diff
@@ -1,6 +1,8 @@
 
 module WebSocketPolyfill {
   requires org.graalvm.polyglot;
+  requires io.helidon.webclient;
+  requires io.helidon.webclient.websocket;
```
Comment (JaroslavTulach) [inline src/main/java/module-info.java:5]: Doubled `requires`!?

File: src/main/java/org/apidesign/polyfill/websocket/Main.java [hunk 2]
```diff
@@ -1,61 +1,53 @@
...
         var path = "/all-y-websocket.js";
-        var demo = WebSocket.class.getResource(path);
+        var demo = Main.class.getResource(path);
         if (demo == null) {
             throw new IOException("Cannot find " + path);
         }
         var commonJsRoot = new File(demo.toURI()).getParent();
+
+        HostAccess hostAccess = HostAccess.newBuilder(HostAccess.EXPLICIT)
+                .allowArrayAccess(true)
```
Comment (JaroslavTulach) [inline src/main/java/org/apidesign/polyfill/websocket/Main.java:27]: Using `HostAccess.EXPLICIT` is good - increases security.

File: src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js [hunk 3]
```diff
@@ -1,24 +1,221 @@
...
+        WebSocket.CONNECTING = 0;
+        WebSocket.OPEN = 1;
+        WebSocket.CLOSING = 2;
+        WebSocket.CLOSED = 3;
+
+        //
+        // Properties
+        //
+
+        Object.defineProperty(WebSocket.prototype, "url", {
```
Comment (JaroslavTulach) [inline src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js:44]: Using `defineProperty`! Why? To control `enumerable`, etc.?

File: src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js [hunk 4]
```diff
@@ -1,24 +1,221 @@
...
+            } else {
+                jvm('web-socket-close', this.connection_, code, reason);
+            }
+        }
+
+        //
+        // EventTarget
+        //
+
+        WebSocket.prototype.addEventListener = function (type, listener) {
```
Comment (JaroslavTulach) [inline src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js:149]: Not using `defineProperty`! Why? Why do we use `defineProperty` once and not in this case?

File: src/test/java/org/apidesign/polyfill/CryptoPolyfillTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        executor = Executors.newSingleThreadExecutor();
+        var crypto = new CryptoPolyfill();
+        var b = Context.newBuilder("js");
+
+        var chromePort = Integer.getInteger("inspectPort", -1);
+        if (chromePort > 0) {
+            b.option("inspect", ":" + chromePort);
+        }
+
+        context = CompletableFuture
```
Comment (JaroslavTulach) [inline src/test/java/org/apidesign/polyfill/CryptoPolyfillTest.java:35]: This feels like a bit too complicated initialization for end users.

**Review by JaroslavTulach — APPROVED** [review]
One question. Few notes.

File: src/main/java/module-info.java (hunk 1 above)
Comment (4e6) [reply src/main/java/module-info.java:5]: The same as for `org.helidon.webserver`

File: src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js (hunk 3 above)
Comment (4e6) [reply src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js:44]: Just to make the field read-only

File: src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js (hunk 4 above)
Comment (4e6) [reply src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js:149]: I only used `defineProperty` for fields to make them read-only

File: src/test/java/org/apidesign/polyfill/CryptoPolyfillTest.java (hunk 5 above)
Comment (4e6) [reply src/test/java/org/apidesign/polyfill/CryptoPolyfillTest.java:35]: yes, it feels quite verbose

File: src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js (hunk 4 above)
Comment (JaroslavTulach) [reply src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js:149]: Why are some fields read only and some are not?

Comment (4e6) [reply src/main/resources/org/apidesign/polyfill/websocket/websocket-polyfill.js:149]: In this case it's a function, not a field

**Pull request merged**